closure/templates/test/ExampleModule.java

```diff
@@ -0,0 +1,94 @@
+// Copyright 2019 The Closure Rules Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package test;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.inject.AbstractModule;
+import com.google.inject.multibindings.Multibinder;
+import com.google.template.soy.data.SoyValue;
+import com.google.template.soy.data.restricted.StringData;
+import com.google.template.soy.exprtree.Operator;
+import com.google.template.soy.jssrc.restricted.JsExpr;
+import com.google.template.soy.jssrc.restricted.SoyJsSrcFunction;
+import com.google.template.soy.jssrc.restricted.SoyJsSrcPrintDirective;
+import com.google.template.soy.shared.restricted.SoyFunction;
+import com.google.template.soy.shared.restricted.SoyJavaFunction;
+import com.google.template.soy.shared.restricted.SoyJavaPrintDirective;
+import com.google.template.soy.shared.restricted.SoyPrintDirective;
+import java.util.List;
+import java.util.Set;
+
+public class ExampleModule extends AbstractModule {
+  /** {@inheritDoc} */
+  @Override
+  protected void configure() {
+    bindFunctions(Multibinder.newSetBinder(binder(), SoyFunction.class));
+    bindDirectives(Multibinder.newSetBinder(binder(), SoyPrintDirective.class));
+  }
+
+  private void bindFunctions(Multibinder<SoyFunction> fns) {
+    fns.addBinding().to(ToLowerFunction.class);
+  }
+
+  private void bindDirectives(Multibinder<SoyPrintDirective> dirs) {
+    dirs.addBinding().to(ToUpperDirective.class);
+  }
+
+  public static class ToLowerFunction implements SoyJavaFunction, SoyJsSrcFunction {
+    @Override
+    public String getName() {
+      return "toLower";
+    }
+
+    @Override
+    public Set<Integer> getValidArgsSizes() {
+      return ImmutableSet.of(1);
+    }
+
+    @Override
+    public SoyValue computeForJava(List<SoyValue> args) {
+      return StringData.forValue(args.get(0).stringValue().toLowerCase());
+    }
+
+    @Override
+    public JsExpr computeForJsSrc(List<JsExpr> args) {
+      JsExpr arg = args.get(0);
+      String exprText = "(" + arg.getText() + ").toLowerCase()";
+      return new JsExpr(exprText, Operator.NOT_EQUAL.getPrecedence());
+    }
+  }
+
+  public static class ToUpperDirective implements SoyJavaPrintDirective, SoyJsSrcPrintDirective {
```

Comment:
It seems like (custom) print directives are [deprecated](https://github.com/google/closure-templates/blob/master/documentation/dev/plugins.md#custom-print-directives-deprecated-and-not-recommended), so I'd rather not use them in the example.

Reply:
Blarg, I didn't realize that. I wish the dev team would value backwards compatibility a bit more. It's just more busywork for the poor suckers using print directives happily, in order to match someone's idea of a cleaner design. </rant> 

Thanks for pointing it out, removed.